core/fetcher.go

```diff
@@ -161,73 +172,81 @@ func (f *BlockFetcher) ValidatorSet(ctx context.Context, height int64) (*types.V
 	return validatorSet, nil
 }
 
-// SubscribeNewBlockEvent subscribes to new block events from Core, returning
-// a new block event channel on success.
-func (f *BlockFetcher) SubscribeNewBlockEvent(ctx context.Context) (<-chan types.EventDataSignedBlock, error) {
-	if f.isListeningForBlocks.Load() {
-		return nil, fmt.Errorf("already subscribed to new blocks")
-	}
-	ctx, cancel := context.WithCancel(ctx)
-	f.cancel = cancel
-	f.doneCh = make(chan struct{})
-	f.isListeningForBlocks.Store(true)
-
-	subscription, err := f.client.SubscribeNewHeights(ctx, &coregrpc.SubscribeNewHeightsRequest{})
-	if err != nil {
-		close(f.doneCh)
-		f.isListeningForBlocks.Store(false)
-		return nil, err
-	}
+func (f *BlockFetcher) runSubscriber() (chan types.EventDataSignedBlock, error) {
+	signedBlockCh := make(chan types.EventDataSignedBlock, 1)
 
-	log.Debug("created a subscription. Start listening for new blocks...")
-	signedBlockCh := make(chan types.EventDataSignedBlock)
 	go func() {
 		defer close(f.doneCh)
 		defer close(signedBlockCh)
-		defer func() { f.isListeningForBlocks.Store(false) }()
 		for {
 			select {
-			case <-ctx.Done():
+			case <-f.ctx.Done():
+				return
+			default:
+			}
```

Comment:
Is it necessary? Looks like next call (SubscribeNewHeights) can exit on canceled context

Reply:
This is the idea. It can be canceled only if context was canceled(in `listener.Stop`)